### PR TITLE
Week6

### PR DESCRIPTION
**Pull request by alyoshenka**
Week 6 submission. Not all routes are implemented, but some are started.

I messed up and branched from week5, instead of main. It made more sense to make the PR into week 5 in this case so that there are fewer changed files to look through.

There are still some extraneous comments and console logs because the assignment is not fully complete.

### CODE REVIEW

**Comment by alyoshenka** [comment]
w6 assignments completed

File: service/app.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+app.use(cors())
+app.use(express.json())
+app.use(exampleMiddleware)
+
+
+// routes
+
+// mount router
+app.use('/', router)
```
Comment (panzerama) [inline service/app.js:31]: :100:

File: service/models/User.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import mongoose from 'mongoose'
+
+const UserSchema = new mongoose.Schema({
+    firstName: String,
+    lastName: String,
+    userId: mongoose.Types.ObjectId
```
Comment (panzerama) [inline service/models/User.js:6]: We don't need a separate userId here. mongo will add an id for us.

**Review by panzerama — COMMENTED** [review]
This is only a partial review, more notes incoming.

File: service/routes/router.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    .post(createCard)
+
+router.route('/users')
+    .post(createUser)
+
+router.route('/users/:id')
+    .put(updateUserById)
+    .delete(deleteUserById)
+
+// module.exports.router = router;
```
Comment (panzerama) [inline service/routes/router.js:49]: Remove unnecessary comment.

File: service/routes/router.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import express from 'express'
+const router = express.Router()
```
Comment (panzerama) [inline service/routes/router.js:2]: Using a router is a great idea, but it's not one I'm going to cover much going forward (for the sake of simplicity). You've used it correctly here, just be aware that you probably won't see it much beyond this.

File: service/routes/routes.js [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                res.status(404).send(`cannot find card ${req.params.id}`)
+            }
+        } else {
+            res.status(404).send(`cannot find deck with card ${req.params.id}`)
+        }
+
+    } catch (err) {
+        res.status(400).send(err.name + err.message)
+    }
+
```
Comment (panzerama) [inline service/routes/routes.js:81]: Watch trailing whitespace

File: service/routes/routes.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+const updateCardById = async(req, res, next) => {
+    const data = req.body
+    try {
+        const deck = await Deck.findOne({
+            'cards._id': req.params.id
+        })
+        if (deck) {
+            const card = deck.cards.find(card => card._id = req.params.id)
```
Comment (panzerama) [inline service/routes/routes.js:37]: deck.cards has an `id` method that you can use  with req.params.id. Check out https://mongoosejs.com/docs/subdocs.html#finding-a-subdocument

File: service/routes/routes.js [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+            if (card) {
+                card.frontImage = data.frontImage ? data.frontImage : card.frontImage
```
Comment (panzerama) [inline service/routes/routes.js:39]: Good use of ternary operator. You can reduce this even further with the [Nullish Coalescing Operator, ??](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator)
```suggestion
                card.frontImage = data.frontImage ?? card.frontImage
```

**Pull request merged**